compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -940,7 +940,7 @@ impl<'a, 'tcx> FindInferSourceVisitor<'a, 'tcx> {
             //
             // See the `need_type_info/issue-103053.rs` test for
             // a example.
-            if !matches!(path.res, Res::Def(DefKind::TyAlias, _)) => {
+            if !matches!(path.res, Res::Def(DefKind::TyAlias | DefKind::AssocTy, _)) => {
```

Comment:
Since we're explicitly looking for an adt below and no other types, can this just be `matches!(Res::Def(DefKind::Struct | DefKind::Enum | DefKind::Union, _))`?

Reply:
Yep, I think this is good advice because it can prevent more similar issues.